File: Regression/Ver 2 - UC with random walk drift term/Matlab/lik_fcn.m
function val = lik_fcn(prmtr,y,T,START,prior)

  %Transform hyperparamters to impose constraints
     
  prmtr = trans(prmtr);
    
  phi_y1 = prmtr(1);
  phi_yh = prmtr(2);
  phi_h1 = prmtr(3);
  phi_hy = prmtr(4);
  
  sig_nyy = prmtr(5)^2; % s.e. of HP permanent component
  sig_nhh = prmtr(6)^2; % s.e. of credit permanent component
  sig_eyy = prmtr(7)^2; % s.e. of the HP AR component
  sig_ehh = prmtr(8)^2; % s.e. of the credit AR component
  sig_nynh = prmtr(9)*sqrt(sig_nyy*sig_nhh);
  sig_eyeh = prmtr(10)*sqrt(sig_eyy*sig_ehh);
  
  sig_wyy = prmtr(11);
  sig_whh = prmtr(12);

    F = [1,1,0,0,0,0; %Transition matrix
         0,1,0,0,0,0;
         0,0,phi_y1,0,0,phi_yh;
         0,0,0,1,1,0;
         0,0,0,0,1,0;
         0,0,phi_hy,0,0,phi_h1];

    Fstar = [phi_y1,phi_yh;
            phi_hy,phi_h1]; %Transition matrix of I(0) part];
    
    H = [1,0,1,0,0,0; %Measurement equation
        0,0,0,1,0,1];

    Q = [sig_nyy,0,0,sig_nynh,0,0; %Cov matrix
        0,sig_wyy,0,0,0,0;
        0, 0,sig_eyy,0, 0, sig_eyeh;
        sig_nynh, 0, 0, sig_nhh, 0, 0;
        0,0,0,0,sig_whh,0;
        0, 0,sig_eyeh, 0, 0, sig_ehh];

    Qstar = [sig_eyy,sig_eyeh; %Cov matrix of I(0) part
            sig_eyeh,sig_ehh];

    A = [0;0];

    beta_ll = [prior(1),0,0,prior(2),0,0]'; %Starting values

    vecQstar = reshape(Qstar,[numel(Qstar),1]);
    vecP_ll = inv(eye(4) - kron(Fstar,Fstar))*vecQstar;
    
    %Var matrix of initial state vector
    P_ll = [prior(3),0,0,prior(7),0,0;
            0,prior(4),0,0,0,0;
            0,0,vecP_ll(1,1),0,0,vecP_ll(2,1);
            prior(7),0,0,prior(5),0,0;
            0,0,0,0,prior(6),0;
            0,0,vecP_ll(3,1),0,0,vecP_ll(4,1)];
                
    lik_mat = zeros(T,1);
    
    %j_iter = 1;
    
    for j_iter = 1:T
        beta_tl  =  F*beta_ll;
        P_tl  =  F*P_ll*F' + Q;

        vt = y(j_iter,1:2)' - H*beta_tl - A; %Prediction error

        ft =  H*P_tl*H'; %Variance of forecast error

        beta_tt = beta_tl + P_tl*H'*inv(ft)*vt;
        P_tt = P_tl - P_tl*H'*inv(ft)*H*P_tl;

        lik_mat(j_iter,1) = 0.5*log(((2*pi)^2)*det(ft)) + 0.5*vt'*inv(ft)*vt;

        beta_ll = beta_tt;
        P_ll = P_tt;
    end

    val = sum(lik_mat(START:T));

end

File: Data Collection/Codes/Ver 2/State Space/lik_fcn.R
lik_fcn <- function(prmtr){
  
  prmtr = trans(prmtr)
  
  phi_y1 = prmtr[1]
  phi_y2 = prmtr[2]
  phi_c1 = prmtr[3]
  phi_c2 = prmtr[4]
  mu = prmtr[5]
  sig_yy = prmtr[6]^2 # s.e. of income AR component
  sig_cc = prmtr[7]^2 # s.e. of consumption AR component
  sig_vv = prmtr[8]^2 # s.e. of the random walk component
  sig_yc = prmtr[9]*sqrt(sig_yy*sig_cc)
  sig_yv = prmtr[10]*sqrt(sig_yy*sig_vv)
  sig_cv = prmtr[11]*sqrt(sig_cc*sig_vv)
  cbar = prmtr[12]
  gamma = prmtr[13]
  
  F = matrix(0,5,5) # Transition matrix
  F[1,] = c(phi_y1,phi_y2,0,0,0)
  F[2,] = c(1,0,0,0,0)
  F[3,] = c(0,0,phi_c1,phi_c2,0)
  F[4,] = c(0,0,1,0,0)
  F[5,] = c(0,0,0,0,1)
  
  Fstar = F[-5,-5] # Transition matrix of I(0) part
  
  muvec = matrix(c(0,0,0,0,mu),5,1) # Drift vector
  
  H = matrix(0,2,5)
  H[1,] = c(1,0,0,0,1)
  H[2,] = c(0,0,1,0,gamma)
  
  Q = matrix(0,5,5) # Cov matrix
  Q[1,] = c(sig_yy,0,sig_yc,0,sig_yv)
  Q[3,] = c(sig_yc,0,sig_cc,0,sig_cv)
  Q[5,] = c(sig_yv,0,sig_cv,0,sig_vv)
  
  Qstar = Q[-5,-5] # Cov matrix of I(0) part
                  # I(0) -> stationary so no trend components
  
  A = matrix(c(0,cbar),2,1)
  
  beta_ll = matrix(c(0,0,0,0,947.5),5,1) # Starting values
  
  vecQstar = matrix(Qstar,ncol = 1)
  vecP_ll = solve(diag(16) - Fstar%x%Fstar)%*%vecQstar
  
  # Var matrix of initial state vector
  P_ll = matrix(0,5,5)
  P_ll[1,] = c(vecP_ll[1,1],0,vecP_ll[3,1],0,0)
  P_ll[3,] = c(vecP_ll[9,1],0,vecP_ll[11,1],0,0)
  P_ll[5,] = c(0,0,0,0,prior)
  
  lik_mat = matrix(0,T,1)
  
  ##Below are steps for update filter
  
  for(j_iter in 1:T){
    beta_tl = muvec + F%*%beta_ll
    P_tl = F%*%P_ll%*%t(F) + Q
    
    vt = matrix(c(y[j_iter,1:2]),2,1) - H%*%beta_tl - A # Prediction error
    ft = H%*%P_tl%*%t(H) # Variance of forecast error
    
    beta_tt = beta_tl + P_tl%*%t(H)%*%solve(ft)%*%vt
    P_tt = P_tl - P_tl%*%t(H)%*%solve(ft)%*%H%*%P_tl
    
    lik_mat[j_iter,1] = 0.5*log(((2*pi)^2)*det(ft)) + 0.5*t(vt)%*%solve(ft)%*%vt
    
    beta_ll = beta_tt
    P_ll = P_tt
  }
  
  val = sum(lik_mat[START:T])
  
  return(val)
}